### PR TITLE
The first-order hyperdoctrine of monomorphisms

### PR DESCRIPTION
**Pull request by nmvdw**
This PR contains:
- some small refactorings to make constructors of fiberwise limits/colimits more useful
- a construction of complete Heyting algebra from posets
- a construction of the first-order hyperdoctrine of monomorphisms

### CODE REVIEW

File: UniMath/CategoryTheory/DisplayedCats/Examples/HValuedPredicates.v [hunk 1]
```diff
@@ -201,8 +201,8 @@ Section HValuedSets.
          apply cha_le_top).
     - abstract
         (cbn ;
-         intros X Y f ;
-         apply idpath).
+         intros X Y f x ;
+         apply cha_le_refl).
```
Comment (benediktahrens) [inline UniMath/CategoryTheory/DisplayedCats/Examples/HValuedPredicates.v:205]: What is the purpose of this change?

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+ in a category. In standard texts on category theory, subobjects of `y` are defined to be
+ monomorphisms `x --> y`. However, this does not lead to the right notion of equality
+ between subobjects. More specifically, two subobjects `x₁ --> y` and `x₂ --> y` should be
+ equal if we have an isomorphism `x₁ --> x₂` making the triangle commute. For this reason,
+ one usually defines the collection of subobjects of `y` as a quotient of the collection of
+ monomorphisms `x --> y`. Note that this is how subobjects are defined in the file
+ `Subobjects.v`.
+
+ If one works with univalent categories, then monomorphisms have the right notion of equality.
+ In addition, the type of monomorphisms into objects `y` automatically form a set. For this
```
Comment (benediktahrens) [inline UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:20]: What does this sentence mean precisely?

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+Require Import UniMath.CategoryTheory.DisplayedCats.Fiber.
+
+Local Open Scope cat.
+
+Definition isaset_disp_ob
+           {C : category}
+           (D : disp_univalent_category C)
+           (H : locally_propositional D)
+           (x : C)
+  : isaset (D x).
```
Comment (benediktahrens) [inline UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:57]: Is this the right file for the definition (lemma?)?

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v [hunk 4]
```diff
@@ -0,0 +1,558 @@
+(*******************************************************************************************
+
+ The displayed category of monomorphisms
+
+ In this file, we define the displayed category of monomorphisms. The definition is very
+ similar to how the arrow category is defined, which is used to define the codomain functor.
+ We also establish several basic facts, namely we characterize isomorphisms and we prove that
+ this displayed category is univalent if the category is univalent.
+
```
Comment (benediktahrens) [inline UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:9]: Would it be a good idea to define this displayed category via the Sigma-construction, with the arrow (displayed) category in between?

File: UniMath/CategoryTheory/DisplayedCats/MonoCodomain/Inclusion.v [hunk 5]
```diff
@@ -0,0 +1,173 @@
+(**************************************************************************************
+
+ The inclusion of monomorphisms into morphisms
+
+ We construct the displayed functor over the identity from the displayed category of
+ monomorphisms to the displayed category of morphisms, and we instantiate it to obtain
+ fiber functors. We also show that an object in the slice category lies in the slice
+ category of monomorphisms if an only if the map to the terminal object is a monomorphism
+ as well.
```
Comment (benediktahrens) [inline UniMath/CategoryTheory/DisplayedCats/MonoCodomain/Inclusion.v:9]: Would this be easier/automatic if monos were defined via the Sigma-construction?

File: UniMath/OrderTheory/Lattice/Examples/Sieves.v [hunk 6]
```diff
@@ -0,0 +1,914 @@
...
+ Complete Heyting Algebras of Sieves
+
+ We consider two ways to construct a Heyting algebra from a poset `X`.
+
+ The first way is via sieves. A sieve in a poset is a downwards closed set. The collection of
+ all sieves then forms a Heyting algebra. The order is given by subset, the lattice operations
+ are given by union and intersection, and the supremum of a collection of sieves is also given
+ by the union.
+
+ For the other construction, we assume that we have a coverage of `X`. A coverage on a poset
```
Comment (benediktahrens) [inline UniMath/OrderTheory/Lattice/Examples/Sieves.v:12]: Naïve question: I have never heard of a "coverage" of something, but only of a "cover" of something. Is "coverage" standard here?

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 2 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:20]: I want to mention the following:
- Subobjects of objects `y` in a strict category are defined to be a quotient of the type of monomorphisms into `y`
- Subobjects of objects `y` in a univalent category are defined to be monomorphisms into `y`

For univalent categories, we don't need to take a quotient.

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 4 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:9]: Good point. Can I do these changes (together with moving `isaset_disp_ob`) in a later PR?

File: UniMath/OrderTheory/Lattice/Examples/Sieves.v (hunk 6 above)
Comment (nmvdw) [reply UniMath/OrderTheory/Lattice/Examples/Sieves.v:12]: I took the terminology from the book 'Intuitionistic Set Theory' by Bell (see page 107 in https://publish.uwo.ca/~jbell/INTBOOK(Repaired).pdf). This definition is quite similar to Grothendieck topologies. I also haven't seen the terminology elsewhere.

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 3 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:57]: Nope. I forgot to move it.

File: UniMath/CategoryTheory/DisplayedCats/Examples/HValuedPredicates.v (hunk 1 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/Examples/HValuedPredicates.v:205]: I changed `make_fiberwise_terminal_locally_propositional`.

The goal is about the stability under substitution. In an earlier version I made `make_fiberwise_terminal_locally_propositional` so that stability was proven by giving an equality. However, it becomes easier and nicer to use if one only has to provide a morphism. As a consequence, I had to give a morphism (`cha_le_refl`) rather than an equality (`idpath`).

File: UniMath/CategoryTheory/DisplayedCats/MonoCodomain/Inclusion.v (hunk 5 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/MonoCodomain/Inclusion.v:9]: It would help if we introduce the following construction:

Given
- a category `C`
- a displayed category `D` over `C`
- a displayed category `E` over `∫ D`
This gives rise to a displayed category `∑ E` over `C`. Then we also have a displayed functor `∑ E --> D` over the identity of `C`. It sends every object `(y_1 , y_2)` over `x` where `y_1 : D x` and `y_2 : E(x , y_1)` to `y_1` over `x`.

As far as I know, we don't have that yet in UniMath.

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 2 above)
Comment (benediktahrens) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:20]: It only forms a set when looking at strict cats. But strictness is not mentioned, is it? Am I missing something?

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 4 above)
Comment (benediktahrens) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:9]: Of course!

File: UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v (hunk 2 above)
Comment (nmvdw) [reply UniMath/CategoryTheory/DisplayedCats/Examples/MonoCodomain.v:20]: For univalent categories `C`, the type `∑ (x : C), Monic x y` is in fact an `hSet`. This is proven in the statement `isaset_subobject_univ_cat`. It is essential for this proof that we talk about monomorphisms, because that is what guarantees that isomorphisms are unique. More specifically, there is at most 1 isomorphism that makes the following diagram commutes

Screenshot 2024-07-15 at 22 53 18

because the involved arrows to `y` are monic.
